### PR TITLE
Implement new Index Column

### LINKED ISSUE #4724: Add better API to blocktree for checking blob dispositions
#### Problem
See this comment: https://github.com/solana-labs/solana/pull/4653#discussion_r295009562

It should be easy to check the disposition of blobs within Blocktree. This is useful in general and will facilitate #4133

#### Proposed Solution
* Implement an Index column similar to the one in the above PR for use in checking blob dispositions in Blocktree. Expose it to clients of Blocktree so that they have an easy way to use it.
* Remove presence tracking functionality from the `ErasureMeta` struct, use the index to do disposition checking.

### PR DESCRIPTION
**Pull request by mark-solana**
#### Problem
See #4724 
#### Summary of Changes
* Implement new index column for disposition checking
* remove disposition checking functions from the `ErasureMeta` column.

Fixes #4724

### CODE REVIEW

File: core/src/blocktree/meta.rs [hunk 1]
```diff
@@ -27,6 +27,115 @@ pub struct SlotMeta {
...
+    pub slot: u64,
+    data: DataIndex,
+    coding: CodingIndex,
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq)]
+pub struct DataIndex {
+    slot: u64,
+    /// Map representing presence/absence of data blobs
+    index: BTreeMap<u64, bool>,
```
Comment (rob-solana) [inline core/src/blocktree/meta.rs]: could this be a HashSet?

Comment (mark-solana) [reply core/src/blocktree/meta.rs]: Yes, I'll change it.

**Comment by pgarg66** [comment]
@rob-solana any concerns if we merge this PR?

File: core/src/erasure.rs [hunk 2]
```diff
@@ -138,23 +138,25 @@ impl Session {
 
             if n < NUM_DATA {
                 let mut blob = Blob::new(&blocks[n]);
+                blob.meta.size = blob.data_size() as usize;
```
Comment (rob-solana) [inline core/src/erasure.rs]: @pgarg66 can you prove to yourself these are the right steps with size?

File: core/src/packet.rs [hunk 3]
```diff
@@ -515,10 +515,10 @@ impl Blob {
     }
 
     pub fn data(&self) -> &[u8] {
-        &self.data[BLOB_HEADER_SIZE..]
+        &self.data[BLOB_HEADER_SIZE..BLOB_HEADER_SIZE + BLOB_DATA_SIZE]
     }
     pub fn data_mut(&mut self) -> &mut [u8] {
-        &mut self.data[BLOB_HEADER_SIZE..]
+        &mut self.data[BLOB_HEADER_SIZE..BLOB_HEADER_SIZE + BLOB_DATA_SIZE]
```
Comment (rob-solana) [inline core/src/packet.rs]: are most uses of data_mut() gone with this PR?

Comment (pgarg66) [reply core/src/packet.rs]: There are two places (serialize) where data_mut is still called. But it has definitely reduced.

File: core/src/erasure.rs (hunk 2 above)
Comment (pgarg66) [reply core/src/erasure.rs]: It seems ok to me. The following is my analysis:

In old code it's calling `blob.set_size()` with value `size - BLOB_HEADER_SIZE`.
```
    pub fn set_size(&mut self, size: usize) {
        let new_size = size + BLOB_HEADER_SIZE;
        self.meta.size = new_size;
        self.set_data_size(new_size as u64);
    }
```

`set_size` adds back the `BLOB_HEADER_SIZE` and updates the meta. In addition, it calls `set_data_size()`. But that's not needed, as we are reading the size from blob itself. So it should already be correct in the blob header.

**Pull request merged**